File: libs/shared-rules-engine/src/lib/rule-engine.ts
/**
 * Generic, composable rule engine.
 *
 * A `Rule` is a named predicate over some input that returns a `RuleResult`.
 * A `RuleEngine` runs a set of rules and aggregates the outcome. Both are
 * generic over the input type so each workflow can bring its own domain object
 * (a Proposta, a Cliente, ...) while reusing the same engine.
 */
export interface RuleResult {
  passed: boolean;
  /** Human-readable explanation, required when a rule fails. */
  reason?: string;
}

export interface Rule<TInput> {
  name: string;
  evaluate(input: TInput): RuleResult | Promise<RuleResult>;
}

export interface RuleOutcome {
  passed: boolean;
  results: Array<{ rule: string } & RuleResult>;
  /** Names of the rules that failed. */
  failures: string[];
}

export interface RuleEngine<TInput> {
  add(rule: Rule<TInput>): RuleEngine<TInput>;
  run(input: TInput): Promise<RuleOutcome>;
}

export function createRuleEngine<TInput>(
  initialRules: Array<Rule<TInput>> = []
): RuleEngine<TInput> {
  const rules: Array<Rule<TInput>> = [...initialRules];

  const engine: RuleEngine<TInput> = {
    add(rule) {
      rules.push(rule);
      return engine;
    },
    async run(input) {
      const results = await Promise.all(
        rules.map(async (rule) => ({
          rule: rule.name,
          ...(await rule.evaluate(input)),
        }))
      );
      const failures = results.filter((r) => !r.passed).map((r) => r.rule);
      return { passed: failures.length === 0, results, failures };
    },
  };

  return engine;
}

File: libs/shared-rules-engine/RuleEngine.cs
namespace Acme.Shared.RulesEngine;

public sealed record RuleResult(bool Passed, string? Reason = null);

/// <summary>A named predicate over some input.</summary>
public interface IRule<in TInput>
{
    string Name { get; }
    RuleResult Evaluate(TInput input);
}

public sealed record EvaluatedRule(string Rule, bool Passed, string? Reason);

public sealed record RuleOutcome(
    bool Passed,
    IReadOnlyList<EvaluatedRule> Results,
    IReadOnlyList<string> Failures);

/// <summary>
/// Generic, composable rule engine. Runs a set of rules over an input and
/// aggregates the outcome. Generic over the input type so each workflow can
/// bring its own domain object while reusing the same engine.
/// </summary>
public sealed class RuleEngine<TInput>
{
    private readonly List<IRule<TInput>> _rules = new();

    public RuleEngine<TInput> Add(IRule<TInput> rule)
    {
        _rules.Add(rule);
        return this;
    }

    public RuleOutcome Run(TInput input)
    {
        var results = _rules
            .Select(rule =>
            {
                var result = rule.Evaluate(input);
                return new EvaluatedRule(rule.Name, result.Passed, result.Reason);
            })
            .ToList();

        var failures = results.Where(r => !r.Passed).Select(r => r.Rule).ToList();
        return new RuleOutcome(failures.Count == 0, results, failures);
    }
}
